server/channels/api4/channel_test.go

```diff
@@ -1834,7 +1897,8 @@ func TestGetChannelsForTeamForUser(t *testing.T) {
 			require.True(t, f, "missing a channel")
 		}
 
-		channels, resp, _ = client.GetChannelsForTeamForUser(context.Background(), th.BasicTeam.Id, th.BasicUser.Id, false, resp.Etag)
+		channels, resp, err = client.GetChannelsForTeamForUser(context.Background(), th.BasicTeam.Id, th.BasicUser.Id, false, resp.Etag)
+		require.Error(t, err, "an error is expected as the server didn't return any data")
```

Comment:
If an etag is used, the server should not return data. This checks validates the logic.

Reply:
I guess this is an unfortunate effect of us building `Client4` largely for our own testing. We want to be able to test ETags, but to work, ETags sort of require a client-side cache to serve up old results which it doesn't have. Without that, you'd have to return nil results for both the data and the error which feels wrong even if it makes sense with it's used

Reply:
> The error then comes from the json decoder, because unmarshaling from an empty reader fails.

Aha, thanks Ben!

Happy to merge this for now with a comment to this discussion for posterity. But I think as a first step, we can simply add some etag handling to return a special error to indicate the lack of etag handling, rather than a json marshalling error. That seems outright wrong.

Reply:
Created https://mattermost.atlassian.net/browse/MM-64633